File: bindings/python/examples/smoke.py
#!/usr/bin/env python3
"""Python smoke test, mirrors examples/smoke.cpp and rust/examples/smoke.rs."""
from nyxstone_tricore_gcc import LabelDefinition, NyxstoneTricoreGCC


def main() -> None:
    nx = NyxstoneTricoreGCC()
    base = 0x80000000

    blob = b""
    for src in [
        "nop",
        "ret",
        "mov %d4, %d5",
        "add %d4, %d5, %d6",
        "movh %d4, 0x1234",
        "start:\n nop\n j here\nhere:\n ret\n",
        ".byte 0x11, 0x22, 0x33, 0x44",
        ".word 0xdeadbeef",
    ]:
        b = nx.assemble(src, address=base + len(blob))
        print(f"--- {src}\n    [{len(b)} bytes]: {b.hex(' ')}")
        blob += b

    print("\n--- external label (j ext, ext = base + 8) ---")
    ext = nx.assemble("nop\n nop\n nop\n j ext\n",
                      address=base,
                      labels=[LabelDefinition("ext", base + 8)])
    print(f"    [{len(ext)} bytes]: {ext.hex(' ')}")

    print("\n--- assemble_with_relocs (gcc/gas -r equivalent) ---")
    rel_bytes, relocs = nx.assemble_with_relocs(
        "nop\n j ext\n",
        address=0x1000,
        labels=[LabelDefinition("ext", 0x2000)])
    print(f"    bytes ({len(rel_bytes)}): {rel_bytes.hex(' ')}")
    for r in relocs:
        print(f"    reloc: off=0x{r.offset:x} type={r.relocation_type} "
              f"sym={r.symbol.name}(@0x{r.symbol.address:x}) addend={r.addend}")

    print("\n--- disassemble_to_instructions ---")
    for ins in nx.disassemble_to_instructions(blob, address=base):
        print(f"    0x{ins.address:08x}  [{ins.bytes.hex(' ')}]  {ins.assembly}")

    print("\n--- disassemble (text, first 3) ---")
    print(nx.disassemble(blob, address=base, count=3), end="")


if __name__ == "__main__":
    main()

File: bindings/rust-ipc/examples/smoke.rs
// Smoke test for the Rust binding, mirrors examples/smoke.cpp.
use nyxstone_tricore_gcc_ipc::{LabelDefinition, NyxstoneTricoreGCC};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let nx = NyxstoneTricoreGCC::new()?;
    let base: u64 = 0x80000000;

    let mut all = Vec::new();
    for src in [
        "nop",
        "ret",
        "mov %d4, %d5",
        "add %d4, %d5, %d6",
        "movh %d4, 0x1234",
        "start:\n nop\n j here\nhere:\n ret\n",
        ".byte 0x11, 0x22, 0x33, 0x44",
        ".word 0xdeadbeef",
    ] {
        let bytes = nx.assemble(src, base + all.len() as u64, &[])?;
        let hex: Vec<String> = bytes.iter().map(|b| format!("{b:02x}")).collect();
        println!("--- {src}\n    [{} bytes]: {}", bytes.len(), hex.join(" "));
        all.extend_from_slice(&bytes);
    }

    println!("\n--- external label (j ext, ext = base + 8) ---");
    let labels = [LabelDefinition::new("ext", base + 8)];
    let ext = nx.assemble("nop\n nop\n nop\n j ext\n", base, &labels)?;
    let hex: Vec<String> = ext.iter().map(|b| format!("{b:02x}")).collect();
    println!("    [{} bytes]: {}", ext.len(), hex.join(" "));

    println!("\n--- assemble_with_relocs (gcc/gas -r equivalent) ---");
    let labels = [LabelDefinition::new("ext", 0x2000)];
    let (rel_bytes, relocs) = nx.assemble_with_relocs(
        "nop\n j ext\n", 0x1000, &labels)?;
    let hex: Vec<String> = rel_bytes.iter().map(|b| format!("{b:02x}")).collect();
    println!("    bytes ({}) : {}", rel_bytes.len(), hex.join(" "));
    for r in &relocs {
        println!("    reloc: off=0x{:x} type={} sym={}({:#x}) addend={:?}",
                 r.offset, r.relocation_type, r.symbol.name,
                 r.symbol.address, r.addend);
    }

    println!("\n--- disassemble_to_instructions ---");
    let insns = nx.disassemble_to_instructions(&all, base, 0)?;
    for ins in &insns {
        let hex: Vec<String> = ins.bytes.iter().map(|b| format!("{b:02x}")).collect();
        println!("    0x{:08x}  [{}]  {}", ins.address, hex.join(" "), ins.assembly);
    }

    println!("\n--- disassemble (text, first 3) ---");
    print!("{}", nx.disassemble(&all, base, 3)?);
    Ok(())
}
